File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid: return 0

        def sink(grid,i,j):
            if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] == "0":
                return 0

            grid[i][j] = "0"
            # up
            sink(grid,i - 1,j)
            # down
            sink(grid,i + 1,j)
            # left
            sink(grid,i,j - 1)
            #right
            sink(grid,i, j + 1)

            return 1


        count = 0

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1":
                    count += sink(grid,i,j)

        
        return count


        # SC -> O(N * M)
        # TC -> O(N * M)
        

File: 200-number-of-islands/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int count = 0;

        for(int i = 0; i < grid.length;i++){
            for(int j = 0;  j < grid[i].length;j++){
                if(grid[i][j] == '1'){ 
                    count+=sinkIsland(i,j,grid);
                }
            }
        }
        
        return count;
    }

    public int sinkIsland(int i , int j, char[][] grid){
        if(i < 0 || i >= grid.length || j < 0 || j >= grid[i].length || grid[i][j] == '0') return 0;

        grid[i][j] = '0';
        sinkIsland(i+1,j,grid);
        sinkIsland(i-1,j,grid);
        sinkIsland(i,j+1,grid);
        sinkIsland(i,j-1,grid);

        
        return 1;
    }
}

// SC -> O(1)
// TC -> O(N*M)